src/main/java/com/conveyal/analysis/models/SpatialResource.java

```diff
@@ -0,0 +1,106 @@
+package com.conveyal.analysis.models;
+
+import com.conveyal.analysis.AnalysisServerException;
+import com.conveyal.analysis.UserPermissions;
+import com.conveyal.analysis.spatial.FeatureSummary;
+import com.conveyal.analysis.spatial.SpatialAttribute;
+import com.conveyal.file.FileStorageFormat;
+import com.conveyal.file.FileStorageKey;
+import com.conveyal.r5.util.ShapefileReader;
+import org.apache.commons.fileupload.FileItem;
+import org.apache.commons.io.FilenameUtils;
+import org.locationtech.jts.geom.Envelope;
+import org.opengis.referencing.FactoryException;
+import org.opengis.referencing.operation.TransformException;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static com.conveyal.file.FileCategory.RESOURCES;
+import static com.conveyal.r5.analyst.Grid.checkWgsEnvelopeSize;
+
+/**
+ * Record of a spatial source file (e.g. shapefile, CSV) that has been validated and can be processed into specific
+ * Conveyal formats (e.g. grids and other spatial layers). Eventually a general Resource class could be extended by
+ * SpatialResource, GtfsResource, and OsmResource?
+ */
+public class SpatialResource extends BaseModel {
```

Comment:
Do we really want a specifically SpatialResource, or a more general Resource with the sourceFormat implying that it's spatial? I guess this one needs geometry type and attribute descriptors, so it's specifically spatial. I'm wondering though where our GTFS and OSM file records fit in.

Reply:
I've gone in circles on this. The original concept of resources would eventually include all user uploads. I then started developing this as standalone spatial database management, and by the end realized (as reflected in the Javadoc highlighted) that a common Resource superclass could make sense.